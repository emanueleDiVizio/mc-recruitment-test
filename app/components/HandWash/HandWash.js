import React, { useState, useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Text, StyleSheet, TouchableOpacity } from 'react-native'
import Icon from 'react-native-vector-icons/FontAwesome5'

import useInterval from '../../hooks/useInterval'
import PropTypes from 'prop-types'
import config from '../../config'
import handWash from '../../state/handWashSlice'

const HandWash = ({
  timerDelay = config.defaultTimerDelay,
  timerTimeout = config.defaultTimerTimeoutInSeconds,
}) => {
  const [timerCount, setTimerCount] = useState(0)
  const [delay, setDelay] = useState(null)
  const [isWashing, setIsWashing] = useState(false)

  const dispatch = useDispatch()

  const onUserPressButton = () => {
    setIsWashing(true)
    setDelay(timerDelay)
  }

  const onTimeOut = () => {
    setDelay(null)
    dispatch(handWash.actions.addHandWash())
    setTimerCount(0)
    setIsWashing(false)
  }

  useEffect(() => {
    if (timerCount === timerTimeout) {
      onTimeOut()
    }
  }, [timerCount])

  useInterval(() => {
    setTimerCount(timerCount + 1)
  }, delay)

  return (
    <TouchableOpacity
      style={styles.circle}
      onPress={onUserPressButton}
      testID="timer-button"
    >
      {isWashing ? (
        <React.Fragment>
          <Text style={styles.countText}>{timerCount}</Text>
          <Text style={styles.secondsText}>Seconds</Text>
        </React.Fragment>
      ) : (
        <React.Fragment>
          <Icon
            name="sign-language"
            size={60}
            color="white"
            style={styles.icon}
          />
          <Text style={styles.washText}>Wash</Text>
        </React.Fragment>
      )}
    </TouchableOpacity>
  )
}

HandWash.propTypes = {
  timerDelay: PropTypes.number,
  timerTimeout: PropTypes.number,
}

const styles = StyleSheet.create({
  circle: {
    width: 200,
    height: 200,
    borderRadius: 200 / 2,
    backgroundColor: config.colors.accent,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
  },
  washText: {
    color: 'white',
    fontSize: 21,
  },
  icon: {
    marginBottom: 24,
  },
  secondsText: {
    color: 'white',
    fontSize: 21,
  },
  countText: {
    color: 'white',
    fontSize: 60,
    marginBottom: 24,
  },
})

export default HandWash
