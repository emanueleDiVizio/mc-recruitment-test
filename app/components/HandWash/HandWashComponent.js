import React, { useState, useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { View, Text, Button } from 'react-native'
import useInterval from '../../hooks/useInterval'
import PropTypes from 'prop-types'

import handWash from '../../state/handWashSlice'

const HandWashComponent = ({ defaultDelay = 1000, defaultTimeout = 20 }) => {
  const [timerCount, setTimerCount] = useState(0)
  const [delay, setDelay] = useState(null)
  const [handWashCompleted, setHandWashCompleted] = useState(false)

  const dispatch = useDispatch()

  const onUserPressButton = () => {
    setDelay(defaultDelay)
  }

  const onTimeOut = () => {
    setDelay(null)
    setHandWashCompleted(true)
  }

  const onHandWashCompleted = () => {
    dispatch(handWash.actions.addHandWash())
    setTimerCount(0)
    setHandWashCompleted(false)
  }

  useEffect(() => {
    if (timerCount === defaultTimeout) {
      onTimeOut()
    }
  }, [timerCount])

  useEffect(() => {
    if (handWashCompleted) {
      onHandWashCompleted()
    }
  }, [handWashCompleted])

  useInterval(() => {
    setTimerCount(timerCount + 1)
  }, delay)

  return (
    <View>
      <Text>Please press the button below to start the timer</Text>
      <Button
        onPress={onUserPressButton}
        testID="timer-button"
        title="Timer"
        disabled={timerCount > 0}
      />
      <Text testID="timer-counter">{timerCount}</Text>
    </View>
  )
}

HandWashComponent.propTypes = {
  defaultDelay: PropTypes.number,
  defaultTimeout: PropTypes.number,
}

export default HandWashComponent
