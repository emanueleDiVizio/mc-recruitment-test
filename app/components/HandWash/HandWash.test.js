import React from 'react'
import HandWash from './HandWash'
import { fireEvent, render, act, wait } from '@testing-library/react-native'
import * as redux from 'react-redux'
import handWash from '../../state/handWashSlice'

describe('HandWash', () => {
  let dispatch

  beforeEach(() => {
    dispatch = jest.fn()
    redux.useDispatch = () => dispatch
  })

  it('initializes correctly', () => {
    const { baseElement } = render(<HandWash />)

    expect(baseElement).toMatchSnapshot()
  })

  it('starts the timer when button is pressed and updates with elapsed time', async () => {
    const { baseElement, getByTestId, getByText } = render(<HandWash />)

    const button = getByTestId('timer-button')
    fireEvent.press(button)

    act(() => jest.advanceTimersByTime(10000))

    await wait(() => expect(getByText('11')).toBeTruthy())

    expect(baseElement).toMatchSnapshot()
  })

  it('dispatches `addHandWash` when timer is complete ', async () => {
    const { getByTestId } = render(<HandWash />)

    const button = getByTestId('timer-button')
    fireEvent.press(button)

    act(() => jest.advanceTimersByTime(22000))

    expect(dispatch).toHaveBeenCalledWith(handWash.actions.addHandWash())
  })
})
