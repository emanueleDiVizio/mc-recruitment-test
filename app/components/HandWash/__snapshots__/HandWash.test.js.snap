// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandWash initializes correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#00A1D2",
          "borderRadius": 100,
          "flexDirection": "column",
          "height": 200,
          "justifyContent": "center",
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": -2,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 16,
          "width": 200,
        }
      }
      testID="timer-button"
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 60,
            },
            Object {
              "marginBottom": 24,
            },
            Object {
              "fontFamily": "FontAwesome5Free-Solid",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {
              "fontWeight": "900",
            },
          ]
        }
      >
        ïŠ§
      </Text>
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 21,
          }
        }
      >
        Wash
      </Text>
    </>
  </View>
</View>
`;
